opt: Add pattern code generation to Optgen

andy-kimball opened the pull request:
The Optgen language already supports pattern matching and replacement
syntax. This PR adds support for generating the code that actually
does the pattern matching and replacement over opt memo expression
trees.

Release note: None

andy-kimball committed:
opt: Add pattern code generation to Optgen

The Optgen language already supports pattern matching and replacement
syntax. This PR adds support for generating the code that actually
does the pattern matching and replacement over opt memo expression
trees.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22454)
<!-- Reviewable:end -->

RaduBerinde commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> :100: 

Very nice! Feel free to ignore the nits, the only thing I would like to understand better is the `negate` handling.

---

Review status: 0 of 11 files reviewed at latest revision, 13 unresolved discussions, some commit checks broke.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 120 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4kjAYI15oymnn8IIkc:-L4kjAYI15oymnn8IIkd:b-b4pcxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6823afb22435723de12072cbe142800b5ab13e5/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L120)):*
> ```Go
> 	g.w.writeIndent("return _group\n")
> 
> 	g.w.unnest(g.w.nesting-1, "}\n")
> ```

This only works because we know the initial nesting level to be 1, right? A comment should explain.

Maybe a better interface would be to have a `marker()` function and then pass the marker to a `unnestToMarker` method.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4kdl-noQJGu2Pj6sl3:-L4kdl-noQJGu2Pj6sl4:b-spkt4p) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6823afb22435723de12072cbe142800b5ab13e5/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L121)):*
> ```Go
> 
> 	g.w.unnest(g.w.nesting-1, "}\n")
> 	g.w.writeIndent("\n")
> ```

[nit] Why `writeIndent`? Btw maybe we should have a `newLine()` or `blankLine()` alias for `write("\n")`

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 84 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l5Ej5Z6Nkwbh0Ur6J:-L4l5Ej5Z6Nkwbh0Ur6K:b-uf9ioo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L84)):*
> ```Go
> 		}
> 
> 		for _, rule := range rules {
> ```

[nit] reads easier if this is inside the `if` below

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 110 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l8rYYYPoRTdnxmsgK:-L4l8rYYYPoRTdnxmsgL:bp9zcu8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L110)):*
> ```Go
> 	for index, matchArg := range rule.Match.Args {
> 		fieldName := unTitle(string(define.Fields[index].Name))
> 		g.genMatch(matchArg, fieldName, false)
> ```

false /* negate */

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 157 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l2p7Qnzcf7EJyarNo:-L4l2p7Qnzcf7EJyarNp:b-jxs63t) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L157)):*
> ```Go
> 		g.genMatchArgs(match, contextName, negate)
> 		return
> 	}
> ```

[nit] this could be a `case m := match.(type)`

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 239 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l7tKJnVM74jZmqlvk:-L4l7tKJnVM74jZmqlvl:b6kr8lc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L239)):*
> ```Go
> 	}
> 
> 	if negate && len(match.Args) != 0 {
> ```

I'm a little confused here. We are also passing `negate` to `gen*MatchArgs` above, which will negate the condition (at least in some cases..). Aren't we negating it twice?

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 252 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4lE0_M40VaI_Z5ujik:-L4lE0_M40VaI_Z5ujil:bs29o4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L252)):*
> ```Go
> // strongly-typed expression structs (e.g. selectExpr, innerJoinExpr, etc).
> func (g *factoryGen) genConstantMatchArgs(
> 	match *lang.MatchExpr, opName string, contextName string, negate bool,
> ```

I think I figured it out - what `negate` means here is slightly different - for `genMatchArgs` it's an instruction to negate the condition, in this case it's just a note that the condition will be negated. Is this right? This deserves some explanation, and maybe a different name for the flag here

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 259 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l9m5kh9br1BwBt1tt:-L4l9m5kh9br1BwBt1tu:b-cxsmlo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L259)):*
> ```Go
> 
> 	if negate && len(match.Args) == 0 {
> 		g.w.nest("if %s == nil {\n", varName)
> ```

It's clearer if we just `return` here. Also, a comment would be useful to explain the logic. In particular, what is the meaning of having no args? Is it that it matches any non-nil expression? In that case I don't know why this condition isn't always `!=` (given that `genMatchArgs` will negate the result).

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 266 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l8g8sc1J3Fxo4pxpM:-L4l8g8sc1J3Fxo4pxpN:bn2q44p) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L266)):*
> ```Go
> 	for index, matchArg := range match.Args {
> 		fieldName := string(g.compiled.LookupDefine(opName).Fields[index].Name)
> 		g.genMatch(matchArg, fmt.Sprintf("%s.%s()", varName, fieldName), false)
> ```

`false /* negate */`. Also, a comment to explain the negate handling would be useful.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 296 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4lFqpUqNFsZL2RxXih:-L4lFqpUqNFsZL2RxXii:b-lzevu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L296)):*
> ```Go
> 	}
> 
> 	if negate && len(match.Args) == 0 {
> ```

same comments as above

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 308 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l8mkglzfDQ6SVO4z1:-L4l8mkglzfDQ6SVO4z2:b-qkuzsu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L308)):*
> ```Go
> 
> 		for index, matchArg := range match.Args {
> 			g.genMatch(matchArg, fmt.Sprintf("%s.ChildGroup(%d)", exprName, index), false)
> ```

`false /* negate */` (and comment explaining)

---

*[pkg/sql/opt/optgen/cmd/optgen/uniquifier.go, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l3hftvSKNDUZV09e4:-L4l3hftvSKNDUZV09e5:bb8js63) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/uniquifier.go#L25)):*
> ```Go
> // added names.
> type uniquifier struct {
> 	unique map[string]bool
> ```

[nit] can be `struct{}` instead of `bool`. Also, maybe rename to `known` or `seen` or similar (`unique[str]` is misleading, if `unique[str] = true` then `str` is actually not unique).

---

*[pkg/sql/opt/optgen/cmd/optgen/testdata/factory, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l4JIMqUxFxkvPP-OW:-L4l4JIMqUxFxkvPP-OX:brq4rkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/testdata/factory#L189)):*
> ```
> 	right opt.GroupID,
> ) opt.GroupID {
> 	_eqExpr := makeEqExpr(left, right)
> ```

just curious, what is the convention with `_`? Everything other than arguments has one? Is this to avoid collisions with arg names ? (though in that case, it feels that the args should be the ones with `_`)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-:-L4l7W2f4UxMIf-432xY:b-ji5la9)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
I added a lot of comments. You can see if you think they help in clearing up the `negate` questions.

---

Review status: 0 of 12 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 120 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4kjAYI15oymnn8IIkc:-L4lb6Fxgdjb0b6Lvt6I:b44q73e) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6823afb22435723de12072cbe142800b5ab13e5/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L120)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This only works because we know the initial nesting level to be 1, right? A comment should explain.

Maybe a better interface would be to have a `marker()` function and then pass the marker to a `unnestToMarker` method.
</blockquote></details>

Yeah, definitely better. I added a `marker()` function.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4kdl-noQJGu2Pj6sl3:-L4lhR0EscD2Lh3-DsKG:bw144mk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6823afb22435723de12072cbe142800b5ab13e5/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L121)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Why `writeIndent`? Btw maybe we should have a `newLine()` or `blankLine()` alias for `write("\n")`
</blockquote></details>

Good idea, I added a newLine() function.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 84 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l5Ej5Z6Nkwbh0Ur6J:-L4lP0ejkUJEcj25UBn4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L84)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] reads easier if this is inside the `if` below
</blockquote></details>

Done.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 110 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l8rYYYPoRTdnxmsgK:-L4lb-BbqXp1V4P74VdG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L110)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

false /* negate */
</blockquote></details>

Done.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 157 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l2p7Qnzcf7EJyarNo:-L4ljdVL7VyTwGZmis1v:bguhsuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L157)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] this could be a `case m := match.(type)`
</blockquote></details>

I can't remember why I didn't use a type switch when I wrote this code in opttoy. I switched this function and others to use a type switch instead. 

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 252 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4lE0_M40VaI_Z5ujik:-L4lnPF32mQEQQPFKth2:bcs2p0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L252)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think I figured it out - what `negate` means here is slightly different - for `genMatchArgs` it's an instruction to negate the condition, in this case it's just a note that the condition will be negated. Is this right? This deserves some explanation, and maybe a different name for the flag here
</blockquote></details>

I renamed the flag to `noMatch`, and added a lot of additional commentary around it in various places.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 259 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l9m5kh9br1BwBt1tt:-L4m0i_NoOposvIammc3:b-e06jds) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L259)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It's clearer if we just `return` here. Also, a comment would be useful to explain the logic. In particular, what is the meaning of having no args? Is it that it matches any non-nil expression? In that case I don't know why this condition isn't always `!=` (given that `genMatchArgs` will negate the result).
</blockquote></details>

Take a look at my new commentary and see if that helps clear this up.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 266 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l8g8sc1J3Fxo4pxpM:-L4m6_TVpMN6vXHDazaL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L266)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`false /* negate */`. Also, a comment to explain the negate handling would be useful.
</blockquote></details>

Done.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 296 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4lFqpUqNFsZL2RxXih:-L4m6_vIYE_Ul1511Yv8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L296)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

same comments as above
</blockquote></details>

Done.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 308 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l8mkglzfDQ6SVO4z1:-L4m6aLA_AYb_KiHx8V0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L308)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`false /* negate */` (and comment explaining)
</blockquote></details>

Done.

---

*[pkg/sql/opt/optgen/cmd/optgen/uniquifier.go, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l3hftvSKNDUZV09e4:-L4lkmKrvtiFJT5FMciG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/uniquifier.go#L25)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] can be `struct{}` instead of `bool`. Also, maybe rename to `known` or `seen` or similar (`unique[str]` is misleading, if `unique[str] = true` then `str` is actually not unique).
</blockquote></details>

Done.

---

*[pkg/sql/opt/optgen/cmd/optgen/testdata/factory, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l4JIMqUxFxkvPP-OW:-L4lkCdgXZMo_hPLimTn:bozwvrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/testdata/factory#L189)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

just curious, what is the convention with `_`? Everything other than arguments has one? Is this to avoid collisions with arg names ? (though in that case, it feels that the args should be the ones with `_`)
</blockquote></details>

If it's a user-defined name that I pull from a pattern, then I don't prefix with `_`. If it's a compiler-generated name, then I do. The reason is to avoid collisions. If we want to get fancier, we could add all the user-defined names to the uniquifier and then use that to avoid collisions instead. But until we do that, I like using the cleaner syntax (i.e. no `_`) for user-defined names and reserving the uglier syntax for compiler-generated names.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-:-L4m6hPQQVIEoxyO6g-O:b-gnl8qz)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 1 of 3 files at r3.
Review status: 1 of 12 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 128 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4m2ER6WM6uwb9vK9WL:-L4m2ER6WM6uwb9vK9WM:bd3aiyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L128)):*
> ```Go
> // The code generation strategy is fairly simple: matchers are generated as a
> // series of nested "if" statements. Within the final "if" statement, the
> // replacement expression is created. Because matchers are often not nested
> ```

Did you mean matchers *are* often nested?

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4m2jtL0CFthXy5bYeQ:-L4m2jtL0CFthXy5bYeR:bqo2i51) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L136)):*
> ```Go
> // against. For example:
> //
> //   for i, _listArg := range _f.mem.lookupList(projections) {
> ```

Nice example!

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 217 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4m7i3gbquwWPa-Po-x:-L4m7i3gbquwWPa-Po-y:b-x8m74q) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf37db1a66fe6d5291932f36b4e2afe51b32aa21/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L217)):*
> ```Go
> 	match *lang.MatchExpr, contextName string, noMatch bool,
> ) {
> 	// The name/child matcher can match multiple parts of the context
> ```

Nice comment - thanks!

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 242 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4m7yfMMDiAngmwtWYV:-L4m7yfMMDiAngmwtWYW:bez1g7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf37db1a66fe6d5291932f36b4e2afe51b32aa21/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L242)):*
> ```Go
> 	// "_match = false" flag, and then testing that flag after breaking out of
> 	// the nested if statements:
> 	//   _match := true
> ```

You've swapped the `_match := true` and `_match := false` here compared to below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 12 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 259 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l9m5kh9br1BwBt1tt:-L4mAjvCvgNNATu75HwM:b-mxqwhf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L259)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Take a look at my new commentary and see if that helps clear this up.
</blockquote></details>

We could still `return` here to make it obvious we aren't doing anything else in this case

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 289 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4mAUnMzFdZvD6j_5H9:-L4mAUnMzFdZvD6j_5HA:benmnj4) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf37db1a66fe6d5291932f36b4e2afe51b32aa21/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L289)):*
> ```Go
> // strongly-typed expression structs (e.g. selectExpr, innerJoinExpr, etc).
> func (g *factoryGen) genConstantMatch(
> 	match *lang.MatchExpr, opName string, contextName string, noMatch bool,
> ```

This is still unclear to me - what does `noMatch` mean for this particular function? Because if we have `Args`, the flag doesn't do anything.

Oh, I see what's going on. We handle the `noMatch && len(match.Args) == 0` differently in multiple pieces of code - we don't generate the `_match = false` stuff at all in this case. This is pretty tricky. I think the code would be much clearer to still use the `_match = false` stuff and not pass `noMatch` here (but the generated code will be messier). Alternatively, we could document that this function supports `noMatch = true` only when there are no args, and pass true only in the no-args case.

I wonder if things would be cleaner if instead of nested ifs we generated something like `if a && b && c`.

---

*[pkg/sql/opt/optgen/cmd/optgen/testdata/factory, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l4JIMqUxFxkvPP-OW:-L4m9mPJseUQtxn1TCPC:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/testdata/factory#L189)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

If it's a user-defined name that I pull from a pattern, then I don't prefix with `_`. If it's a compiler-generated name, then I do. The reason is to avoid collisions. If we want to get fancier, we could add all the user-defined names to the uniquifier and then use that to avoid collisions instead. But until we do that, I like using the cleaner syntax (i.e. no `_`) for user-defined names and reserving the uglier syntax for compiler-generated names.
</blockquote></details>

:thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 11 files at r1, 1 of 2 files at r2, 2 of 3 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optgen/cmd/optgen/testdata/factory, line 281 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4mDY7ZmjytWBlz_bdB:-L4mDY7ZmjytWBlz_bdC:b-7r2bwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf37db1a66fe6d5291932f36b4e2afe51b32aa21/pkg/sql/opt/optgen/cmd/optgen/testdata/factory#L281)):*
> ```
> }
> 
> [NormalizeVarOrder]
> ```

Change name of rule

---

*[pkg/sql/opt/optgen/cmd/optgen/testdata/factory, line 476 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4mEi0K6-j6vfh5jBhO:-L4mEi0K6-j6vfh5jBhP:blbu7cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf37db1a66fe6d5291932f36b4e2afe51b32aa21/pkg/sql/opt/optgen/cmd/optgen/testdata/factory#L476)):*
> ```
> 	}
> 
> 	return _f.onConstruct(_f.mem.memoizeNormExpr((*memoExpr)(&_innerJoinExpr)))
> ```

Even though there are no rules for these methods, should they check for `if !_f.allowOptimizations()`, and then return without using `onConstruct`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-:-L4mGFoD2nmAyailQo_S:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 10 of 12 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 128 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4m2ER6WM6uwb9vK9WL:-L4mFgcufwKyx4YR33Cz:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L128)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Did you mean matchers *are* often nested?
</blockquote></details>

Fixed.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 239 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l7tKJnVM74jZmqlvk:-L4mXCUkFygJlZIrGl6_:bmlktem) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L239)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I'm a little confused here. We are also passing `negate` to `gen*MatchArgs` above, which will negate the condition (at least in some cases..). Aren't we negating it twice?
</blockquote></details>

New logic should address this.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 259 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l9m5kh9br1BwBt1tt:-L4mIAS0Xbc6oIqs2-2y:b-se74sw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L259)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We could still `return` here to make it obvious we aren't doing anything else in this case
</blockquote></details>

I made a different change that removes the need to check for the no-args case.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 242 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4m7yfMMDiAngmwtWYV:-L4mG-ZcVs5CgB5N2244:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf37db1a66fe6d5291932f36b4e2afe51b32aa21/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L242)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

You've swapped the `_match := true` and `_match := false` here compared to below.
</blockquote></details>

Fixed

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 289 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4mAUnMzFdZvD6j_5H9:-L4mGRNd8BRFd727PD__:b1w2drm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf37db1a66fe6d5291932f36b4e2afe51b32aa21/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L289)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This is still unclear to me - what does `noMatch` mean for this particular function? Because if we have `Args`, the flag doesn't do anything.

Oh, I see what's going on. We handle the `noMatch && len(match.Args) == 0` differently in multiple pieces of code - we don't generate the `_match = false` stuff at all in this case. This is pretty tricky. I think the code would be much clearer to still use the `_match = false` stuff and not pass `noMatch` here (but the generated code will be messier). Alternatively, we could document that this function supports `noMatch = true` only when there are no args, and pass true only in the no-args case.

I wonder if things would be cleaner if instead of nested ifs we generated something like `if a && b && c`.
</blockquote></details>

OK, I think I understand. I made a code change that hopefully clarifies things. Take a look.

Re: your last point, the problem is that we need to be able to use statements. Child matchers may contain bindings which need to generate variable declarations and assignments. For example:
```
(Eq $left:(Variable) $right:^(Variable))
```
The left and right expressions can be arbitrarily nested with bindings scattered throughout. I don't see how that could be done with expressions.

---

*[pkg/sql/opt/optgen/cmd/optgen/testdata/factory, line 281 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4mDY7ZmjytWBlz_bdB:-L4mVxjaayQ2p6RfnVYE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf37db1a66fe6d5291932f36b4e2afe51b32aa21/pkg/sql/opt/optgen/cmd/optgen/testdata/factory#L281)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Change name of rule
</blockquote></details>

Done.

---

*[pkg/sql/opt/optgen/cmd/optgen/testdata/factory, line 476 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4mEi0K6-j6vfh5jBhO:-L4mWRmvWMbsvOZVbafF:b-s7tzwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf37db1a66fe6d5291932f36b4e2afe51b32aa21/pkg/sql/opt/optgen/cmd/optgen/testdata/factory#L476)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Even though there are no rules for these methods, should they check for `if !_f.allowOptimizations()`, and then return without using `onConstruct`?
</blockquote></details>

I was planning on having onConstruct check it, but maybe it'd be better to do as you suggest. I'll switch it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
OK, I made some code changes that should make the negation logic easier to understand. Thanks for all the comments. It's useful when I hear what's not easy to understand, as I'm so deep in some of this stuff that I can't always tell what others will say is hard and/or easy to grok.

---

Review status: 10 of 12 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-:-L4mXMLGrNKhHTI_KnjB:bbz5wi3)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 259 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l9m5kh9br1BwBt1tt:-L4mZxvWAy_i0Max_m9C:b-f7386s) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L259)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I made a different change that removes the need to check for the no-args case.
</blockquote></details>

I realize it works so I hate to dwell on this, but it's still unclear what `noMatch` means at the level of this function. In particular passing `noMatch=true` when we have args would now generate bogus code `if x == nil { ... code that uses x.a() ... }`

I'd be ok with just a comment saying that noMatch can only be true if there are no args.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 289 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4mAUnMzFdZvD6j_5H9:-L4mZd_gE9qy5_TySAQU:b7six7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf37db1a66fe6d5291932f36b4e2afe51b32aa21/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L289)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

OK, I think I understand. I made a code change that hopefully clarifies things. Take a look.

Re: your last point, the problem is that we need to be able to use statements. Child matchers may contain bindings which need to generate variable declarations and assignments. For example:
```
(Eq $left:(Variable) $right:^(Variable))
```
The left and right expressions can be arbitrarily nested with bindings scattered throughout. I don't see how that could be done with expressions.
</blockquote></details>

Ah I see, thanks!

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 260 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4mZtOKp9mJ4iIIX6v0:-L4mZtOLd4Hk7CJXCiR0:bslpttv) ([raw file](https://github.com/cockroachdb/cockroach/blob/62a7ed93bebd526e2c50ee5269c7776df81026c2/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L260)):*
> ```Go
> 	// can easily be done by flipping == to/from != in one if statement.
> 	// There's no need to invert execution logic in that case.
> 	invertExecution := noMatch && len(match.Args) != 0
> ```

Nice, this is much easier to understand.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 11 of 12 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 259 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l9m5kh9br1BwBt1tt:-L4me0DyxEjRqpLnrvM4:b-izjc50) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L259)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I realize it works so I hate to dwell on this, but it's still unclear what `noMatch` means at the level of this function. In particular passing `noMatch=true` when we have args would now generate bogus code `if x == nil { ... code that uses x.a() ... }`

I'd be ok with just a comment saying that noMatch can only be true if there are no args.
</blockquote></details>

Yeah, that's a good point. The `genConstantMatch` and `genDynamicMatch` functions are the result of refactoring (ironically for clarity) . This code used to be just inlined in `genMatchNameAndChildren`, in which case we wouldn't have as much of an issue. I'll add an assert in the code under the noMatch condition, and then return early.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 10 of 12 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/opt/optgen/cmd/optgen/factory_gen.go, line 259 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-L4l9m5kh9br1BwBt1tt:-L4mmhKFVqEnh9oSdgrk:b-br0bxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae17fc3a623720cfa431b001cb76bda5c3aaf040/pkg/sql/opt/optgen/cmd/optgen/factory_gen.go#L259)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Yeah, that's a good point. The `genConstantMatch` and `genDynamicMatch` functions are the result of refactoring (ironically for clarity) . This code used to be just inlined in `genMatchNameAndChildren`, in which case we wouldn't have as much of an issue. I'll add an assert in the code under the noMatch condition, and then return early.
</blockquote></details>

Nice, this looks great, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22454#-:-L4mmkdZRTx-athBlVOb:bnfp4nl)*
<!-- Sent from Reviewable.io -->
